Fix split-horizon LB for MCS

nmittler opened the pull request:
Previous logic was broken for the case when only some clusters export service endpoints (via MCS ServiceExport). In this case, it's possible that we can accidentally hit unexported endpoints that share the same network as exported endpoints.

Refactoring the load balancing logic to prefer gateways with the same cluster as the endpoint. Going forward, we should recommend installing east-west gateway in all clusters to support the MCS use case.

Fixes #33526



[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure


Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

nmittler commented on pilot/pkg/networking/util/util.go:
```diff
@@ -480,18 +480,22 @@ func MergeAnyWithAny(dst *any.Any, src *any.Any) (*any.Any, error) {
+	if networkID != "" {
+		addIstioEndpointLabel(metadata, "network", &pstruct.Value{Kind: &pstruct.Value_StringValue{StringValue: networkID.String()}})
+	}
+	// TODO(nmittler): Is there any concern with adding an extra label?
```
@howardjohn any concerns here?

howardjohn commented:
/test benchmark_istio

howardjohn commented on pilot/pkg/networking/util/util.go:
```diff
@@ -480,18 +480,22 @@ func MergeAnyWithAny(dst *any.Any, src *any.Any) (*any.Any, error) {
+	if networkID != "" {
+		addIstioEndpointLabel(metadata, "network", &pstruct.Value{Kind: &pstruct.Value_StringValue{StringValue: networkID.String()}})
+	}
+	// TODO(nmittler): Is there any concern with adding an extra label?
```
Yes. I don't have the numbers off hand but I think even adding one label leads to a big impact on EDS cost (since there are so many endpoints). Why do we need it ?

Actually confused why we have `network` as well.. We should check the benchmark results

howardjohn replied:
Only one way to find out: https://github.com/istio/istio/pull/33717

nmittler replied:
@stevenctl ^^

I suspect the reason for `network` was similar to why I added `cluster` ... some information from `IstioEndpoint` is lost post transformation to the Envoy structs.

stevenctl replied:
We only need it for this in-control-plane logic. If we can fetch it after the fact (find the original istio endpoints) or if we can strip it after this is used that would be good. 

For the "stripping after use" case, we probably want to only add it conditionally, based on MultiNetworkEnabled

nmittler replied:
I've removed `network` and `cluster` from LBEndpoint metadata.

stevenctl commented on pilot/pkg/xds/ep_filters.go:
```diff
@@ -61,73 +60,91 @@ func (b *EndpointBuilder) EndpointsByNetworkFilter(endpoints []*LocLbEndpointsAn
+			// us to more easily spread traffic to the endpoint across multiple network gateways, increasing
+			// reliability of the endpoint. When multiplying, be careful to avoid overflow - clipping the
+			// result at the maximum value for uint32.
+			weight := b.scaleEndpointLBWeight(lbEp, endpointLBWeightScaleFactor)
```
Two comments:
* For the most part the changes here and the problem you're solving could be dealt with simply by filtering the gateways before calculating multiples/performing filtering _but_  it may be easier to understand (took me a while to wrap my head around the original logic)
* I'm not confident a constant scale factor is a good idea. We _should_ be able to calculate one that guarantees divisibility by len(gateways):
   * `factorial(max(possibleGatewayLens))`

nmittler replied:
> For the most part the changes here and the problem you're solving could be dealt with simply by filtering the gateways before calculating multiples/performing filtering but it may be easier to understand (took me a while to wrap my head around the original logic)

Agreed, refactoring the logic wasn't necessary, but I think it is an improvement for understandability and future maintenance as a result.

> I'm not confident a constant scale factor is a good idea. We should be able to calculate one that guarantees divisibility by len(gateways)

Do you have a specific case in mind as to why it might not be a good idea? This is a reasonably hot data path, so I'd like to avoid a lot of extra traversals if possible. Acknowledge that 4 is a bit of a magic number, but it's also reasonably high WRT spreading traffic across gateways on the same network or network+cluster.

stevenctl replied:
The second we hit 5 gateways we can have issues where we don't maintain the original balance (if only slightly). It's the sort of subtle problem that would be hard to diagnose in the future. Systems or vendors that prefer several gateways for reliability could be quick to cause issues. 

The factorial logic isn't very different from our original multiplier calculation, if anything it's slightly quicker and would produce a smaller common factor. 

I don't feel _too_ strongly, but I don't think the cost is high. We can check with benchmarks.

stevenctl commented on pilot/pkg/xds/ep_filters.go:
```diff
@@ -138,6 +155,24 @@ func (b *EndpointBuilder) EndpointsByNetworkFilter(endpoints []*LocLbEndpointsAn
 	return filtered
 }
 
+func (b *EndpointBuilder) selectNetworkGateways(nw network.ID, c cluster.ID) []*model.NetworkGateway {
```
This is a big improvement – can you add a comment explaining the benefits:
* Potentially reduce hops
* MCS case

hzxuzhonghu commented on pilot/pkg/xds/eds.go:
```diff
@@ -315,11 +315,9 @@ func (s *DiscoveryServer) generateEndpoints(b EndpointBuilder) *endpoint.Cluster
 
-	// If networks are set (by default they aren't) apply the Split Horizon
-	// EDS filter on the endpoints
-	if b.push.NetworkManager().IsMultiNetworkEnabled() {
```
nit: I prefer this way

nmittler replied:
I'd rather have this sort of logic bundled into the network filter and always apply it. It makes the call site simpler and pushes all the networking-related decisions down into the filter.

istio-testing commented:
@nmittler: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | dbb531946ba1d9824cb7f24a7d63f687bf393b8f | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33715/lint_istio/1410300092316913664) | `/test lint_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

howardjohn commented:
I would like to review this closely but may not find time today, can you bug me if I don't get to it and am blocking?

nmittler commented:
@howardjohn @stevenctl I think I've addressed all of your comments. I ended up fishing the cluster/network through by passing the original `IstioEndpoint` to the endpoint builder. PTAL

istio-testing commented:
@nmittler: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | a7059fc7dd111c167cc48374e23bece7593de1f0 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33715/unit-tests_istio/1410332755971018752) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@nmittler: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | ecb93a8ea8c2ffe5ccf0a94f9b321ec4f88920c9 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33715/unit-tests_istio/1410338290539696128) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@nmittler: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | 7406a4bb234ed3a77336fb306ee7e34c3734d309 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33715/unit-tests_istio/1410340938043101184) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

nmittler commented:
/retest

istio-testing commented:
@nmittler: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | 7406a4bb234ed3a77336fb306ee7e34c3734d309 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33715/unit-tests_istio/1410340938043101184) | `/test unit-tests_istio`
integ-security-multicluster-tests_istio | 7406a4bb234ed3a77336fb306ee7e34c3734d309 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33715/integ-security-multicluster-tests_istio/1410340944703655936) | `/test integ-security-multicluster-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

nmittler commented:
/retest

stevenctl commented:
/test benchmark_istio

stevenctl commented:
At a glance this looks good, but I am curious how `kb/msg` is going down so much. I haven't reviewed the latest changes, are we dropping some metadata? 

```
name                                          old time/op        new time/op        delta
EndpointGeneration/1/100-8                          1.78ms ± 4%        1.45ms ± 2%  -18.52%  (p=0.008 n=5+5)
EndpointGeneration/10/10-8                          1.22ms ± 2%        1.01ms ± 1%  -17.24%  (p=0.008 n=5+5)
EndpointGeneration/100/10-8                         12.1ms ± 1%        10.1ms ± 1%  -16.22%  (p=0.008 n=5+5)
EndpointGeneration/1000/1-8                         12.5ms ± 2%        10.3ms ± 2%  -17.59%  (p=0.008 n=5+5)
name                                          old alloc/op       new alloc/op       delta
EndpointGeneration/1/100-8                           444kB ± 0%         374kB ± 0%  -15.87%  (p=0.008 n=5+5)
EndpointGeneration/10/10-8                           272kB ± 0%         245kB ± 0%   -9.94%  (p=0.008 n=5+5)
EndpointGeneration/100/10-8                         2.58MB ± 0%        2.33MB ± 0%   -9.47%  (p=0.016 n=4+5)
EndpointGeneration/1000/1-8                         2.57MB ± 0%        2.34MB ± 0%   -9.15%  (p=0.008 n=5+5)
name                                          old allocs/op      new allocs/op      delta
EndpointGeneration/1/100-8                           8.41k ± 0%         6.91k ± 0%  -17.84%  (p=0.008 n=5+5)
EndpointGeneration/10/10-8                           5.08k ± 0%         4.34k ± 0%  -14.59%  (p=0.008 n=5+5)
EndpointGeneration/100/10-8                          46.8k ± 0%         39.8k ± 0%  -15.07%  (p=0.016 n=4+5)
EndpointGeneration/1000/1-8                          46.3k ± 0%         39.3k ± 0%  -15.22%  (p=0.008 n=5+5)
name                                          old kb/msg         new kb/msg         delta
EndpointGeneration/1/100-8                            11.5 ± 0%           9.1 ± 0%  -20.89%  (p=0.008 n=5+5)
EndpointGeneration/10/10-8                            8.53 ± 0%          6.13 ± 0%  -28.14%  (p=0.008 n=5+5)
EndpointGeneration/100/10-8                           83.2 ± 0%          59.2 ± 0%  -28.84%  (p=0.008 n=5+5)
EndpointGeneration/1000/1-8                           83.6 ± 0%          59.6 ± 0%  -28.71%  (p=0.008 n=5+5)
name                                          old resources/msg  new resources/msg  delta
EndpointGeneration/1/100-8                             100 ± 0%           100 ± 0%     ~     (all equal)
EndpointGeneration/10/10-8                            10.0 ± 0%          10.0 ± 0%     ~     (all equal)
EndpointGeneration/100/10-8                           10.0 ± 0%          10.0 ± 0%     ~     (all equal)
EndpointGeneration/1000/1-8                           1.00 ± 0%          1.00 ± 0%     ~     (all equal)
```

howardjohn reviewed (commented):
@stevenctl I believe this incorporates https://github.com/istio/istio/pull/33717

nmittler commented:
Yeah that's right

nmittler commented on pilot/pkg/xds/eds_sh_test.go:
```diff
@@ -296,3 +337,12 @@ func addNetwork(server *xds.FakeDiscoveryServer, id network.ID, network *meshcon
+	for _, lbEp := range eps {
+		addrs = append(addrs, lbEp.GetEndpoint().Address.GetSocketAddress().Address)
+	}
+	sort.Strings(addrs)
```
Discussed offline. Decided that this sort should remain, so that the return values are deterministic. The eds test has a sort as well, which was the main concern here, but it's only used for formatting for the error strings - not validation.

nmittler commented on pilot/pkg/xds/eds_sh_test.go:
```diff
@@ -271,13 +309,16 @@ func initRegistry(server *xds.FakeDiscoveryServer, clusterNum int, gatewaysIP []
 	istioEndpoints := make([]*model.IstioEndpoint, numOfEndpoints)
 	for i := 0; i < numOfEndpoints; i++ {
+		addr := fmt.Sprintf("10.%d.0.%d", networkNum, i+1)
+		fmt.Printf("NM: adding address %v\n", addr)
```
Fixed by https://github.com/istio/istio/pull/33776

hzxuzhonghu commented on pilot/pkg/xds/ep_filters.go:
```diff
@@ -61,73 +61,101 @@ func (b *EndpointBuilder) EndpointsByNetworkFilter(endpoints []*LocLbEndpointsAn
+			remainingWeight := weight
+			for remainingWeight > 0 {
+				// Spread the remaining weight across the gateways.
+				weightPerGateway := remainingWeight / uint32(len(gateways))
```
How about 3/2? How can it reflect the real proportion
It is confusing how the weight is calculated.

nmittler replied:
All we're doing here is spreading the weight for the endpoint across all gateways. We could simply round robin across each gateway, subtracting 1 from the endpoint's weight each time (in essence, that's what this code is doing). The division here is just a way to quickly determine the portion of the endpoint weight that each gateway should receive. This calculation should never be `0`, since we scale each endpoint weight by the maximum gateways per network. But even if it were, we adjust it to `1` below.

hzxuzhonghu commented on pilot/pkg/xds/ep_filters.go:
```diff
@@ -61,73 +61,101 @@ func (b *EndpointBuilder) EndpointsByNetworkFilter(endpoints []*LocLbEndpointsAn
+			istioEndpoint := ep.istioEndpoints[i]
+			epNetwork := istioEndpoint.Network
+			epCluster := istioEndpoint.Locality.ClusterID
+			gateways := b.selectNetworkGateways(epNetwork, epCluster)
```
Are only cluster local endpoints visible to the cluster gateway? How did you make it?

nmittler replied:
The gateway (like all proxies) will only see endpoints for the service that are cluster local or have been exported.
